File: worker/src/routes/covers.ts
/**
 * CORS-friendly cover-image proxy.
 *
 * Why this exists
 * ---------------
 * Tidal's image CDN (`resources.tidal.com`) does **not** send
 * `Access-Control-Allow-Origin` headers. The browser still renders
 * those URLs through `<img>` (no CORS check needed for the display
 * pipeline), but a programmatic `fetch(url)` from the page origin
 * gets rejected outright with a CORS error before the response body
 * is even seen.
 *
 * The offline-save path (`src/lib/offline/streamResolver.ts ::
 * fetchCoverBlob`) needs the actual bytes so it can stash them in
 * IndexedDB and feed `URL.createObjectURL` to `<img>` when the
 * device is offline. Its previous shape tried CORS first and fell
 * back to `mode: 'no-cors'` on rejection. The opaque response works
 * for `<img>` rendering but `.blob()` returns an empty / unreliable
 * blob in some browser/build combinations (especially when the HTTP
 * cache is cold), so saved tracks ended up with `coverBlob:
 * undefined`. While online the UI rendered the network URL and the
 * bug stayed invisible; offline the saved tile fell through to the
 * generic Disc3 / initials placeholder. Reported by the user as
 * "обложки не показываются в плеере и в списке самих треков, только
 * на загруженных плейлистах работает".
 *
 * Fix
 * ---
 * Proxy `resources.tidal.com` (and a small allowlist of related
 * Tidal-owned image hosts) through this endpoint. The worker fetches
 * the upstream image server-side, so there's no CORS check on the
 * client-side fetch, and the global `corsMiddleware` in `index.ts`
 * already attaches `Access-Control-Allow-Origin: *` to every worker
 * response. The client-side `fetchCoverBlob` therefore gets a fully
 * readable Blob with size + mime intact.
 *
 * The endpoint is intentionally unauthenticated for the same reason
 * `/tracks/audio` is: an `<img src="...">` cannot send an
 * Authorization header, and embedding tokens in cover URLs leaks
 * fresh JWTs into Referer / browser history / CF access logs per
 * tile. The hosts allowlist below is the only abuse mitigation we
 * need — it's the same shape and rationale as the audio proxy
 * (`worker/src/routes/tracks.ts :: TIDAL_CDN_ALLOWED`).
 */
import { Hono } from 'hono';
import type { Env, Variables } from '../types/env';

const covers = new Hono<{ Bindings: Env; Variables: Variables }>();

// Hosts the proxy is willing to fetch from. Mirrors the audio
// proxy's allowlist intent: Tidal-owned CDNs only, no wildcard
// CloudFront / Akamai entries that would let an attacker point this
// endpoint at arbitrary public infrastructure.
const COVER_HOSTS_ALLOWED: RegExp[] = [
  /^resources\.tidal\.com$/i,
  /^(.+\.)?tidal\.com$/i,
];

covers.get('/proxy', async (c) => {
  const target = c.req.query('url');
  if (!target) return c.json({ error: 'missing url' }, 400);

  let parsed: URL;
  try {
    parsed = new URL(target);
  } catch {
    return c.json({ error: 'invalid url' }, 400);
  }
  if (parsed.protocol !== 'https:') {
    return c.json({ error: 'https only' }, 400);
  }
  const host = parsed.hostname.toLowerCase();
  if (!COVER_HOSTS_ALLOWED.some((re) => re.test(host))) {
    return c.json({ error: `host not allowed: ${host}` }, 400);
  }

  // Forward an `If-None-Match` so the worker → upstream → 304 short
  // circuit stays intact when the browser already has the cover in
  // its HTTP cache.
  const upstreamHeaders = new Headers();
  const ifNoneMatch = c.req.header('If-None-Match');
  if (ifNoneMatch) upstreamHeaders.set('If-None-Match', ifNoneMatch);

  const upstream = await fetch(target, { headers: upstreamHeaders });

  // Long-cache the proxied image at the edge. Cover URLs include the
  // image-id in the path (`/{uuid}/640x640.jpg`) and the upstream
  // never updates a given path — the only way to "change" a cover is
  // to mint a new image-id. So a 30-day cache is safe and slashes
  // origin traffic for popular tiles.
  const out = new Headers();
  for (const k of [
    'content-type',
    'content-length',
    'etag',
    'last-modified',
  ]) {
    const v = upstream.headers.get(k);
    if (v) out.set(k, v);
  }
  out.set('cache-control', 'public, max-age=2592000, immutable');
  // Mirror the audio proxy's CORS exposure so the client can read
  // content-length for progress reporting if it ever wants to.
  out.set('access-control-expose-headers', 'Content-Length, Content-Type, ETag');

  return new Response(upstream.body, { status: upstream.status, headers: out });
});

export { covers };

File: api-go/internal/routes/covers.go
package routes

import (
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/bratan-corp/bratan-music/api-go/internal/app"
	"github.com/bratan-corp/bratan-music/api-go/internal/httpx"
)

// Ported from worker/src/routes/covers.ts.
//
// CORS-friendly cover-image proxy. Tidal's image CDN
// (resources.tidal.com) does not send
// `Access-Control-Allow-Origin` headers, so a programmatic
// `fetch(url)` from the page origin is rejected with a CORS error
// before the response body is even visible. The offline-save path
// (src/lib/offline/streamResolver.ts :: fetchCoverBlob) needs the
// actual bytes so it can stash them in IndexedDB. Routing the
// request through this endpoint sidesteps the CORS check because
// the global CORS middleware on the API attaches
// `Access-Control-Allow-Origin: *` to every response.
//
// The endpoint is intentionally unauthenticated for the same reason
// the audio proxy is: an `<img src="...">` cannot send an
// Authorization header, and embedding tokens in cover URLs would
// leak fresh JWTs into Referer / browser history / access logs.
// The host allowlist below is the only abuse mitigation we need,
// mirroring the audio proxy's TIDAL_CDN_ALLOWED.

var coverHostsAllowed = []*regexp.Regexp{
	regexp.MustCompile(`(?i)^resources\.tidal\.com$`),
	regexp.MustCompile(`(?i)^(.+\.)?tidal\.com$`),
}

// HTTP client reused across cover-proxy calls. Disable redirect
// auto-follow control — Tidal's image CDN doesn't redirect, but we
// keep the default 10-hop limit which is fine if the policy ever
// changes upstream. A shared client gets us pooled keep-alives, so
// hot covers reuse a single TCP/TLS session.
var coverHTTPClient = &http.Client{
	Timeout: 20 * time.Second,
}

// proxyCover fetches `?url=<upstream>` and streams the response
// back to the caller with long edge-cache headers when the upstream
// returned a 2xx. Non-2xx responses get `cache-control: no-store`
// so a transient upstream 403 / 5xx never gets cached for a month.
func proxyCover(a *app.App) http.HandlerFunc {
	_ = a
	return func(w http.ResponseWriter, r *http.Request) {
		target := r.URL.Query().Get("url")
		if target == "" {
			httpx.Err(w, http.StatusBadRequest, "missing url")
			return
		}
		parsed, err := url.Parse(target)
		if err != nil {
			httpx.Err(w, http.StatusBadRequest, "invalid url")
			return
		}
		if parsed.Scheme != "https" {
			httpx.Err(w, http.StatusBadRequest, "https only")
			return
		}
		host := strings.ToLower(parsed.Hostname())
		allowed := false
		for _, re := range coverHostsAllowed {
			if re.MatchString(host) {
				allowed = true
				break
			}
		}
		if !allowed {
			httpx.Err(w, http.StatusBadRequest, "host not allowed: "+host)
			return
		}

		req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, target, nil)
		if err != nil {
			httpx.Internal(w, err)
			return
		}
		// Forward `If-None-Match` so the upstream 304 short-circuit
		// stays intact when the browser already has the cover in
		// its HTTP cache.
		if inm := r.Header.Get("If-None-Match"); inm != "" {
			req.Header.Set("If-None-Match", inm)
		}

		resp, err := coverHTTPClient.Do(req)
		if err != nil {
			httpx.Err(w, http.StatusBadGateway, "upstream fetch failed")
			return
		}
		defer resp.Body.Close()

		out := w.Header()
		for _, k := range []string{"Content-Type", "Content-Length", "ETag", "Last-Modified"} {
			if v := resp.Header.Get(k); v != "" {
				out.Set(k, v)
			}
		}
		// Only long-cache successful responses. Cover URLs include
		// the image-id in the path (`/{uuid}/640x640.jpg`) and the
		// upstream never updates a given path — the only way to
		// change a cover is to mint a new image-id — so 30 days
		// is safe and slashes origin traffic for popular tiles.
		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			out.Set("Cache-Control", "public, max-age=2592000, immutable")
		} else {
			out.Set("Cache-Control", "no-store")
		}
		// Mirror the audio proxy's CORS exposure so the client can
		// read Content-Length for progress reporting if it ever
		// wants to.
		out.Set("Access-Control-Expose-Headers", "Content-Length, Content-Type, ETag")

		w.WriteHeader(resp.StatusCode)
		// `io.Copy` is fine here — the body is streamed straight to
		// the client without a full in-memory buffer.
		_, _ = io.Copy(w, resp.Body)
	}
}
